tensorflow_probability/python/sts/components/autoregressive_moving_average.py

```diff
@@ -0,0 +1,238 @@
+            level_scale: Scalar (any additional dimensions are treated as batch
+                dimensions) `float` `Tensor` indicating the standard deviation of the
+                transition noise at each step.
+            initial_state_prior: instance of `tfd.MultivariateNormal`
+                representing the prior distribution on latent states.  Must have
+                event shape `[order]`.
+            observation_noise_scale: Scalar (any additional dimensions are
+                treated as batch dimensions) `float` `Tensor` indicating the standard
+                deviation of the observation noise.
+                Default value: 0.
+            name: Python `str` name prefixed to ops created by this class.
+                Default value: "AutoregressiveStateSpaceModel".
+            **linear_gaussian_ssm_kwargs: Optional additional keyword arguments to
+                to the base `tfd.LinearGaussianStateSpaceModel` constructor.
+
+        Notes:
+            This representation is a ARMA(p, p - 1) process where q = p - 1.
+            If q + 1 != p, then either `ar_coefficients` or `ma_coefficients`
+            will be padded with zeros by the required amount to become a
+            ARMA(p, p - 1) process.
+        """
+        parameters = dict(locals())
+        parameters.update(linear_gaussian_ssm_kwargs)
+        del parameters['linear_gaussian_ssm_kwargs']
+        with tf.name_scope(name or 'ARMAStateSpaceModel') as name:
+            # The initial state prior determines the dtype of sampled values.
+            # Other model parameters must have the same dtype.
+            dtype = initial_state_prior.dtype
+
+            ar_coefficients = tf.convert_to_tensor(
+                value=ar_coefficients, name='ar_coefficients', dtype=dtype)
+            ma_coefficients = tf.convert_to_tensor(
+                value=ma_coefficients, name='ma_coefficients', dtype=dtype)
+            level_scale = tf.convert_to_tensor(
+                value=level_scale, name='level_scale', dtype=dtype)
+            observation_noise_scale = tf.convert_to_tensor(
+                value=observation_noise_scale,
+                name='observation_noise_scale', dtype=dtype)
+
+            p = tf.compat.dimension_value(ar_coefficients.shape[-1])
+            q = tf.compat.dimension_value(ma_coefficients.shape[-1])
+
+            if p is None or q is None:
+                raise ValueError('coefficients must have static shape.')
+
+            # if p != q + 1, then pad either of the arguments to ensure
+            # we end up fitting an ARMA(p, p-1) process.
+            if p > q + 1:
+                pad = tf.zeros([p - q - 1], dtype=dtype)
+                ma_coefficients = tf.concat([ma_coefficients, pad], axis=-1)
+            elif q + 1 > p:
+                pad = tf.zeros([q + 1 - p], dtype=dtype)
+                ar_coefficients = tf.concat([ar_coefficients, pad], axis=-1)
```

Comment:
I was getting `dtype` issues with your approach. Conceptually, this code is a little easier for me:
```
def _pad_tensor_with_trailing_zeros(x, num_zeros):
    pad = tf.zeros_like(x, dtype=x.dtype)[..., 0:1]
    pad = tf.repeat(pad, num_zeros, axis=-1)
    return tf.concat([x, pad], axis=-1)
```
So maybe I will propose this as it passes tests. 

Also, as this padding may frustrate users because the event shape of the prior needs to match `max(p, q + 1)`, I think adding:
```
            # ensure the prior's shape matches the padded order
            prior_event_shape = tf.compat.dimension_value(
                initial_state_prior.event_shape[-1])
            if prior_event_shape != self.order:
                raise ValueError('prior event shape needs to match max(p, q + 1). '
                                 f'prior event shape: {prior_event_shape}, '
                                 f'max(p, q + 1): {self.order}.')
```
would be best

Reply:
Ah, right---my suggestion should have included `dtype` in the `tf.zeros([tf.rank(x) - 1, 2], dtype=x.dtype)` term. Your approach also looks correct, so doing it that way is fine by me.

Minor nit: `tf.zeros_like(x, dtype=x.dtype)[..., 0:1]` would be better as `tf.zeros_like(x[..., 0:1])`, since passing `x.dtype` to `zeros_like` is redundant, and doing the slicing inside avoids allocating an unnecessary temporary array of zeros. (though I'm sure the actual performance difference is negligible).